### ISSUE TITLE
question about the loss computaion

### ISSUE THREAD

**Issue by MengHao666** [NVlabs/DiffusionNFT#13]
in your code, you compute the positive_loss and negative_loss with the prediction `x_0`
https://github.com/NVlabs/DiffusionNFT/blob/24af5554898d85e93efa492f42b5e9cdf4156e9b/scripts/train_nft_sd3.py#L947
https://github.com/NVlabs/DiffusionNFT/blob/24af5554898d85e93efa492f42b5e9cdf4156e9b/scripts/train_nft_sd3.py#L955

however, in the description and Pseudocode of your paper
you compute it with prediction "velocity"

<img width="1790" height="1006" alt="Image" src="https://github.com/user-attachments/assets/c443f9ee-73f7-40ff-927e-b54cfe8f46db" />

**Comment by ChenDRAG** [comment]
You can also train with a velocity field. Below is our original code.

With `config.train.normalized_loss` set to False, we perform DiffusionNFT as in the paper.

Training in `x0` field is just a "trick" that helps stabilize training. Please see adaptive weighting part in the paper. (Figure 9 and sec 3.3)

```
                                    assert "flow" in config.loss_style
                                    target = noise - x0
                                    # normalize advantage
                                    normalized_advantages_clip = (advantages_clip / config.train.adv_clip_max) / 2.0 + 0.5
                                    r = torch.clamp(normalized_advantages_clip, 0, 1)
                                    implicit_negative_prediction = forward_prediction
                                    loss_terms["old_deviate"] = torch.mean((forward_prediction - old_prediction)**2).detach()
                                    loss_terms["old_deviate_max"] = torch.max((forward_prediction - old_prediction)**2).detach()
                                    positive_prediction = forward_prediction
                                    if not config.train.normalized_loss:
                                        positive_loss = ((positive_prediction - target)**2).mean(dim=(1, 2, 3))
                                        negative_loss = -((implicit_negative_prediction - target)**2).mean(dim=(1, 2, 3))
                                    else:
                                        assert config.weight_t == 0
                                        x0_prediction = xt - t[:, None, None, None] * positive_prediction
                                        with torch.no_grad():
                                            weight_factor = (torch.abs(x0_prediction.double() - x0.double()).mean(dim=[1, 2, 3], keepdim=True).clip(min=0.00001))
                                        positive_loss = ((x0_prediction - x0)**2/weight_factor).mean(dim=(1, 2, 3))
                                        negative_x0_prediction = xt - t[:, None, None, None] * implicit_negative_prediction
                                        with torch.no_grad():
                                            negative_weight_factor = (torch.abs(negative_x0_prediction.double() - x0.double()).mean(dim=[1, 2, 3], keepdim=True).clip(min=0.00001))
                                        negative_loss = -((negative_x0_prediction - x0)**2/negative_weight_factor).mean(dim=(1, 2, 3))
                                    ori_policy_loss = r * positive_loss  + (1.0-r) * negative_loss * config.negweight
```

**Comment by MengHao666** [comment]
thanks for info !

**Comment by Smadx** [comment]
Hi, I found that the loss in scripts/train_nft_sd3.py line 959 is not the same as what you provided above:
```python
ori_policy_loss = r * positive_loss / config.beta + (1.0 - r) * negative_loss / config.beta # line 959
policy_loss = (ori_policy_loss * config.train.adv_clip_max).mean()
```
Could you please explain why we need to divide by another ``beta`` here and why we need to multiply by ``adv_clip_max`` on the next row?
